### DISCUSSION TITLE
Merge two HeteroData objects

### DISCUSSION THREAD

**Discussion by AdrianBZG** [pyg-team/pytorch_geometric#4492] [Q&A]
Hello,

Is it possible to join two HeteroData objects (two graphs) into a single HeteroData one (single graph)? With the assumption that they are subgraphs of the same big graph (hence same node types, edge types, etc).

Any pointers on how to accomplish this would be appreciated.

Thanks.

**Comment by rusty1s** [comment]
I think you should be able to simply batch them together:
```python
from torch_geometric.data import Batch

data = Batch.from_data_list([data1, data2])
```

**Reply by AdrianBZG** [reply]
Thanks for that, will that handle possible duplicated edges? The the two subgraphs (HeteroData) as subgraphs ontained through NeighborLoader, so they will probably have duplicated edges.

**Reply by rusty1s** [reply]
There won‘t exists any duplicated edges. Both subgraphs are treated independently and won‘t have any connections with each other.

**Reply by AdrianBZG** [reply]
Yes I realised this when I saw the description of how the method works. Is there any way of joining them by "hacking" the internal data structures (merge both attributes of each node type, merge edge_index from all relations) such that the new merged graph will have the nodes that are the same in one subgraph with the same IDs of the second subgraph? Or any suggestions on how to accomplish this?

**Reply by rusty1s** [reply]
This is not easily possible since there is information missing in the subgraphs which nodes overlap in both graphs. The easiest approach would be to simply re-sample data using `loader.neighbor_sampler(node_indices)` for the union of node indices. Another approach is to track the global node indices of both graphs, and have some routine to merge `edge_index`:
```python
data.n_id = torch.arange(data.num_nodes)

loader = NeighborLoader(data, ...)

iter = iter(loader)
data1 = next(iter)
data2 = next(iter)

global_edge_index1 = data1.n_id[data1.edge_index]
global_edge_index2 = data2.n_id[data2.edge_index]

global_edge_index = torch.cat([global_edge_index1, global_edge_index2], dim=-1)

# Move `global_edge_index back to local indices
n_id = torch.cat([data1.n_id, data2.n_id]).unique()
mapping = torch.empty(data.num_nodes, dtype=torch.long)
mapping[n_id] = torch.arange(n_id.numel())
edge_index = mapping[global_edge_index]
```
This is untested but hopefully gets you going.

**Reply by AdrianBZG** [reply]
Thank you for this. In the case of relations between different node types, I should just get the global edge indexes from the corresponding node type n_id, no?

Regarding that, if I have a relation NodeType1 -> Relation1 -> NodeType2, the node indexes for NodeType1 will be in edge_index[0] or [1]?

**Reply by rusty1s** [reply]
In case of `HeteroData`, one would need to do this for every edge type indeed. `edge_index[0]` will correspond to `NodeType1`, `edge_index[1]` will correspond to `NodeType2`.

**Comment by nooblyh** [comment]
@rusty1s  Hello, I have another question on merging HeteroData objects using `Batch.from_data_list()`. If one of the HeteroData objects has more node types and edge index types than the others, these additional node types and edge index types seem to be discarded after the collation. But if I keep every HeteroData object with the same node types and edge index types, and assign an empty tensor to those node types and edge index types that have no values, I can't pass it to the GNN because of the zero shapes. Is there any better way of doing this?

**Reply by nooblyh** [reply]
OK, I guess I just need to pop the empty tensor in the x_dict and edge_index_dict and that will work.

**Reply by rusty1s** [reply]
Glad that this is resolved. I am pretty confident that empty tensors inside `MessagePassing` are properly handled. If you think there is a bug somewhere, please let me know.

**Reply by nooblyh** [reply]
Actually, I find this in HANConv's forward function:
```
# Iterate over node types:
for node_type, x_node in x_dict.items():
    x_node_dict[node_type] = self.proj[node_type](x_node).view(
        -1, H, D)
    out_dict[node_type] = []
```
If x_node is an empty tensor I got an error like `RuntimeError: mat1 and mat2 shapes cannot be multiplied`

**Reply by rusty1s** [reply]
I just pushed a test that confirms that `HANConv` can operate on empty tensors, see https://github.com/pyg-team/pytorch_geometric/pull/4756. Can you please checkout PyG master and see if this already fixes issues on your end?

**Reply by nooblyh** [reply]
Ah, I see what happened. I assigned the "empty tensor" in shape (0,) instead of shape (0, feature_dim) when building the HeteroData object. Thank you so much for your time! :heart:
